### PR TITLE
tap: updated tcl code to c++ code

### PR DESCRIPTION
**Pull request by Salmaafifi98**
Signed-off-by: Salmaafifi98 <salmaafifi98@aucegypt.edu>

### CODE REVIEW

File: src/tap/src/tapcell.cpp [hunk 1]
```diff
@@ -33,41 +33,1306 @@
...
+  filled_sites_.clear();
+}
+
+void Tapcell::cutRows(odb::dbMaster* endcap_master,
+                      vector<odb::dbBox*> blockages,
+                      int& halo_x,
+                      int& halo_y)
+{
+  odb::dbBlock* block = db_->getChip()->getBlock();
+  int rows_count = (block->getRows()).size();
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:163]: drop the extra ()

File: src/tap/src/tapcell.cpp [hunk 2]
```diff
@@ -33,41 +33,1306 @@
...
+  int block_count = blockages.size();
+  int min_row_width;
+  int end_width;
+
+  if (endcap_master != nullptr) {
+    end_width = endcap_master->getWidth();
+    min_row_width = 2 * end_width;
+  } else {
+    min_row_width = 0;
+  }
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:173]: Just

const int min_row_width = endcap_master ? 2 * endcap_master->getWidth() : 0;

File: src/tap/src/tapcell.cpp [hunk 3]
```diff
@@ -33,41 +33,1306 @@
...
+  // Gather rows needing to be cut up front
+  vector<odb::dbRow*> blocked_rows;
+  std::map<string, vector<odb::dbBox*>> row_blockages;
+  string row_name;
+  for (odb::dbBox* blockage : blockages) {
+    for (odb::dbRow* row : block->getRows()) {
+      if (overlaps(blockage, row, halo_x, halo_y)) {
+        row_name = row->getName();
+        if (row_blockages.find(row_name) == row_blockages.end()) {
+          blocked_rows.push_back(row);
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:184]: There is no need to use the row name,  just use dbRow* which is unique and faster.  You can also eliminated blocked_rows as the keys of row_blockages are the rows.

File: src/tap/src/tapcell.cpp [hunk 4]
```diff
@@ -33,41 +33,1306 @@
...
+          vector<odb::dbBox*> row_blockages_bbox;
+          row_blockages_bbox.push_back(blockage);
+          row_blockages.insert(std::pair<string, vector<odb::dbBox*>>(
+              row_name, row_blockages_bbox));
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:188]: row_blockages.insert({row_name, row_blockages_bbox}); should work  (row_name should be row after my above comment).

File: src/tap/src/tapcell.cpp [hunk 5]
```diff
@@ -33,41 +33,1306 @@
...
+        }
+      }
+    }
+  }
+
+  // cut rows around macros
+  for (odb::dbRow* row : blocked_rows) {
+    cutRow(block, row, row_blockages, min_row_width, halo_x, halo_y);
+  }
+  int cut_rows_count = ((block->getRows()).size()) - rows_count;
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:200]: drop ()

File: src/tap/src/tapcell.cpp [hunk 6]
```diff
@@ -33,41 +33,1306 @@
...
+        } else {
+          row_blockages[row_name].push_back(blockage);
+        }
+      }
+    }
+  }
+
+  // cut rows around macros
+  for (odb::dbRow* row : blocked_rows) {
+    cutRow(block, row, row_blockages, min_row_width, halo_x, halo_y);
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:198]: You don't need to pass all of row_blockages, just row_blockages[row]

File: src/tap/src/tapcell.cpp [hunk 7]
```diff
@@ -33,41 +33,1306 @@
...
+  }
+  int cut_rows_count = ((block->getRows()).size()) - rows_count;
+  logger_->info(utl::TAP, 1, "Found {} macro blocks.", block_count);
+  logger_->info(utl::TAP, 2, "Original rows: {}", rows_count);
+  logger_->info(utl::TAP, 3, "Created {} rows for a total of {} rows.", cut_rows_count, cut_rows_count+rows_count);
+}
+
+void Tapcell::cutRow(odb::dbBlock* block,
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:206]: Where is the current row being cut actually removed?

File: src/tap/src/tapcell.cpp [hunk 8]
```diff
@@ -33,41 +33,1306 @@
...
+                           odb::dbMaster* endcap_master,
+                           vector<string>& cnrcap_masters)
+{
+  int start_phy_idx = phy_idx_;
+  odb::dbBlock* block = db_->getChip()->getBlock();
+  odb::dbMaster* cnrcap_nwin_master;
+  odb::dbMaster* cnrcap_nwout_master;
+
+  int endcapwidth = endcap_master->getWidth();
+  int do_corners;
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:279]: is this a bool?

File: src/tap/src/tapcell.cpp [hunk 9]
```diff
@@ -33,41 +33,1306 @@
...
+    odb::Rect row_bb;
+    row->getBBox(row_bb);
+    if (x >= row_bb.xMin() && x <= row_bb.xMax()) {
+      return true;
+    }
+  }
+  return false;
 }
 
+int Tapcell::isXCorner(const int x,
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:408]: 'is' methods are usually boolean predicates.  What do the -1,0,1 return values mean?

File: src/tap/src/tapcell.cpp [hunk 10]
```diff
@@ -33,41 +33,1306 @@
...
+  inst->setLocation(x, y);
+  inst->setPlacementStatus(odb::dbPlacementStatus::LOCKED);
+  inst->setSourceType(odb::dbSourceType::DIST);
+
+  odb::dbBox* inst_bb = inst->getBBox();
+  vector<int> filled_sites_vector;
+  filled_sites_vector.push_back(inst_bb->yMin());
+  filled_sites_vector.push_back(inst_bb->xMin());
+  filled_sites_vector.push_back(inst_bb->xMax());
+  filled_sites_.push_back(filled_sites_vector);
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:1221]: Use a struct not a vector to hold these values

File: src/tap/src/tapcell.cpp [hunk 11]
```diff
@@ -33,41 +33,1306 @@
...
+  int start_phy_idx = phy_idx_;
+  odb::dbBlock* block = db_->getChip()->getBlock();
+  odb::dbMaster* cnrcap_nwin_master;
+  odb::dbMaster* cnrcap_nwout_master;
+
+  int endcapwidth = endcap_master->getWidth();
+  int do_corners;
+  if (cnrcap_masters.size() == 2) {
+    string cnrcap_nwin_master_name = cnrcap_masters[0];
+    string cnrcap_nwout_master_name = cnrcap_masters[1];
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:282]: This looks like it should be a struct rather than a two element vector.

File: src/tap/src/tapcell.cpp [hunk 12]
```diff
@@ -33,41 +33,1306 @@
...
+      subrow->getBBox(row_bb);
+      auto row_ori = subrow->getOrient();
+
+      const int llx = row_bb.xMin();
+      const int lly = row_bb.yMin();
+      const int urx = row_bb.xMax();
+      const int ury = row_bb.yMax();
+
+      odb::dbMaster* masterl;
+      odb::dbMaster* masterr;
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:326]: left_master & right_master are clearer

File: src/tap/src/tapcell.cpp [hunk 13]
```diff
@@ -33,41 +33,1306 @@
...
+    vector<int> x_start_end;
+    x_start_end.push_back(x_start);
+    x_start_end.push_back(x_end);
+    if (row_fills.find(y) == row_fills.end()) {
+      vector<vector<int>> vector2;
+      vector2.push_back(x_start_end);
+      row_fills.insert(std::pair<int, vector<vector<int>>>(y, vector2));
+    } else {
+      row_fills[y].push_back(x_start_end);
+    }
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:477]: I think this can just be 

row_fills[y].push_back(x_start_end);

If y isn't present in row_fills it will be inserted by [] and a vector default constructed.

File: src/tap/src/tapcell.cpp [hunk 14]
```diff
@@ -33,41 +33,1306 @@
...
+  }
+
+  for (std::map<int, vector<vector<int>>>::iterator iter = row_fills.begin();
+       iter != row_fills.end();
+       ++iter) {
+    y = iter->first;
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:483]: Just
for (const auto& [y, ignored] : row_fills) {

File: src/tap/src/tapcell.cpp [hunk 15]
```diff
@@ -33,41 +33,1306 @@
...
+  std::map<std::pair<int, int>, vector<int>> macro_outlines
+      = getMacroOutlines(rows);
+  for (std::map<std::pair<int, int>, vector<int>>::iterator iter
+       = macro_outlines.begin();
+       iter != macro_outlines.end();
+       ++iter) {
+    vector<int> bot_top_row = iter->second;
+    for (int i = 0; i < bot_top_row.size(); i += 2) {
+      bot_top_row[i]--;
+      bot_top_row[i + 1]++;
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:520]: This is hard to follow - please explain?

File: src/tap/src/tapcell.cpp [hunk 16]
```diff
@@ -33,41 +33,1306 @@
...
+      const int site_x = row->getSite()->getWidth();
+      odb::Rect row_bb;
+      row->getBBox(row_bb);
+      int llx = row_bb.xMin();
+      const int lly = row_bb.yMin();
+      const int urx = row_bb.xMax();
+      const int ury = row_bb.yMax();
+      odb::dbOrientType ori = row->getOrient();
+      int offset = 0;
+      int pitch = -1;
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:562]: Please put variables as close to their first point of usage as possible (and in the smallest scope possible).

File: src/tap/src/tapcell.cpp [hunk 17]
```diff
@@ -33,41 +33,1306 @@
...
+    } else {
+      master = tap_nwouttie_master;
+      tb2_master = tap_nwout2_master;
+      tb3_master = tap_nwout3_master;
+    }
+  }
+
+  int tbtiewidth = master->getWidth();
+  int tap3_master_width = tb3_master->getWidth();
+  int tap2_master_width = tb2_master->getWidth();
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:765]: It looks like many variables could be const

File: src/tap/src/tapcell.cpp [hunk 18]
```diff
@@ -33,41 +33,1306 @@
...
+  if ((remaining_distance != 0) && (remaining_distance < tap2_master_width)) {
+    tbtiecount--;
+  }
+  // insert tb tie
+  int x = x_start;
+  for (int n = 0; n < tbtiecount; n++) {
+    if (checkSymmetry(master, ori)) {
+      buildCell(block, master, ori, x, lly, prefix);
+    }
+    x = x + tbtiewidth;
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:779]: +=

File: src/tap/src/tapcell.cpp [hunk 19]
```diff
@@ -33,41 +33,1306 @@
...
+            utl::TAP, 12, "Master {} not found.", cnrcap_nwin_master_name);
+      }
+      cnrcap_nwout_master = db_->findMaster(cnrcap_nwout_master_name.c_str());
+
+      if (cnrcap_nwout_master == nullptr) {
+        logger_->error(
+            utl::TAP, 13, "Master {} not found.", cnrcap_nwout_master_name);
+      }
+      std::pair<int, int> min_max_x;
+      min_max_x = getMinMaxX(rows);
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:302]: What does this do?  It seems like a no-op

File: src/tap/src/tapcell.cpp [hunk 20]
```diff
@@ -33,41 +33,1306 @@
...
+        }
+      }
+    }
+  }
+  int blkgs_cnt = phy_idx_ - start_phy_idx;
+  logger_->info(utl::TAP, 7, "Inserted {} cells near blockages.", blkgs_cnt);
+  return blkgs_cnt;
+}
+
+std::pair<int, int> Tapcell::getMinMaxX(vector<vector<odb::dbRow*>>& rows)
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:1028]: const method

File: src/tap/src/tapcell.cpp [hunk 21]
```diff
@@ -33,41 +33,1306 @@
...
+{
+  int min_x = -1;
+  int max_x = -1;
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:1031]: Just initialize these with numeric_limits<int>::max() and min() and use std::max() min() below without any if statements.

File: src/tap/src/tapcell.cpp [hunk 22]
```diff
@@ -33,41 +33,1306 @@
...
+{
+  std::map<std::pair<int, int>, vector<int>> macro_outlines;
+  std::pair<int, int> min_max_x = getMinMaxX(rows);
+
+  for (int cur_row = 0; cur_row < rows.size(); cur_row++) {
+    vector<odb::dbRow*> subrows = rows[cur_row];
+    int subrow_count = subrows.size();
+
+    int macro0 = -1;
+    int macro1 = -1;
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:1066]: Are there better names?

File: src/tap/src/tapcell.cpp [hunk 23]
```diff
@@ -33,41 +33,1306 @@
...
+        max_x = new_max_x;
+      } else if (max_x < new_max_x) {
+        max_x = new_max_x;
+      }
+    }
+  }
+  return {min_x, max_x};
+}
+
+std::map<std::pair<int, int>, vector<int>> Tapcell::getMacroOutlines(
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:1055]: A comment describing the return value would be helpful.

File: src/tap/src/tapcell.cpp [hunk 24]
```diff
@@ -33,41 +33,1306 @@
...
+    } else {
+      macro_outlines_array[k].insert(
+          macro_outlines_array[k].end(), new_rows.begin(), new_rows.end());
+    }
+  }
+
+  return macro_outlines_array;
+}
+
+// // proc to detect if blockage overlaps with row
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:1135]: Extra //
not a proc anymore

File: src/tap/src/tapcell.cpp [hunk 25]
```diff
@@ -33,41 +33,1306 @@
...
+    }
+  }
+
+  return blockages;
+}
+
+int Tapcell::makeSiteLoc(int x, double site_x, int dirc, int& offset)
+{
+  auto site_int = double(x - offset) / site_x;
+  if (dirc == 1) {
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:1187]: What is dirc and why is it an int?

File: src/tap/src/tapcell.cpp [hunk 26]
```diff
@@ -33,41 +33,1306 @@
...
+    site_int = ceil(site_int);
+  } else {
+    site_int = floor(site_int);
+  }
+  site_int = int(site_int * site_x + offset);
+  return site_int;
+}
+
+void Tapcell::buildCell(odb::dbBlock* block,
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:1196]: makeInstance would be more descriptive

File: src/tap/src/tapcell.cpp [hunk 27]
```diff
@@ -33,41 +33,1306 @@
...
+                        string prefix)
+{
+  if (x < 0 || y < 0) {
+    return;
+  }
+  string name = prefix + std::to_string(phy_idx_);
+  odb::dbInst* inst = odb::dbInst::create(block, master, name.c_str());
+  if (inst == nullptr) {
+    return;
+  }
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:1210]: Is it really ok to return or is this an error of some sort?  It can only really happen with a duplicate name.

File: src/tap/src/tapcell.cpp [hunk 28]
```diff
@@ -33,41 +33,1306 @@
...
+                       start_x,
+                       y,
+                       orient,
+                       direction,
+                       new_row_num_sites,
+                       site_width);
+  }
+}
+
+vector<vector<odb::dbRow*>> Tapcell::organizeRows()
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:1254]: Organize how?

File: src/tap/src/tapcell.cpp [hunk 29]
```diff
@@ -33,41 +33,1306 @@
...
+    odb::Rect rowBB;
+    row->getBBox(rowBB);
+    if (rows_dict.find(rowBB.yMin()) == rows_dict.end()) {
+      vector<odb::dbRow*> rows_vector;
+      rows_vector.push_back(row);
+      rows_dict.insert(
+          std::pair<int, vector<odb::dbRow*>>(rowBB.yMin(), rows_vector));
+    } else {
+      rows_dict[rowBB.yMin()].push_back(row);
+    }
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:1267]: Just
 rows_dict[rowBB.yMin()].push_back(row);

File: src/tap/src/tapcell.cpp [hunk 30]
```diff
@@ -33,41 +33,1306 @@
...
+  string prefix_string(prefix);
+  int removed = 0;
+
+  if (prefix_string.length() == 0) {
+    // if no prefix is given, return 0 instead of having all cells removed
+    return 0;
+  }
+
+  for (odb::dbInst* inst : block->getInsts()) {
+    if (inst->getName().find(prefix_string) != string::npos) {
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:1309]: Are you really intending to find the prefix anywhere in the string?  If so it isn't really a prefix.  If not, use substr to get the string prefix for comparison.

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/tap/src/tapcell.cpp [hunk 31]
```diff
@@ -33,41 +33,1306 @@
...
-void
-Tapcell::init(Tcl_Interp *tcl_interp,
-	   odb::dbDatabase *db)
+void Tapcell::init(odb::dbDatabase* db, utl::Logger* logger)
 {
   db_ = db;
+  logger_ = logger;
+}
+
+void Tapcell::reset()
```
[the remark this answers is not in the document]
Comment (Salmaafifi98) [reply src/tap/src/tapcell.cpp:69]: They have separate 'set' functions; setTapPrefix and setEndcapPrefix. Is that ok?

File: src/tap/src/tapcell.cpp (hunk 3 above)
Comment (Salmaafifi98) [reply src/tap/src/tapcell.cpp:184]: I adjusted the row name and how the values are inserted into the map as advised in the other comments. However if the keys of row_blockages are used instead of blocked_rows, the boundary_macros test fails as it finds differences in the def file. I believe this is because the keys in the map are arranged in ascending order automatically as opposed to the order the rows are being inserted in the blocked_rows vector.

File: src/tap/src/tapcell.cpp (hunk 7 above)
Comment (Salmaafifi98) [reply src/tap/src/tapcell.cpp:206]: At the end of the cutRow function the current row after it is cut is removed by odb::dbRow::destroy(row);

File: src/tap/src/tapcell.cpp (hunk 8 above)
Comment (Salmaafifi98) [reply src/tap/src/tapcell.cpp:279]: Yes. it's adjusted now, thank you.

File: src/tap/src/tapcell.i [hunk 32]
```diff
@@ -34,14 +34,166 @@
...
+  }
+
+  void clear()
+  {
+    getTapcell()->clear();
+  }
+
+  void reset() { getTapcell()->reset(); }
+
+  int make_site_loc(int x, double site_x, int dirc, int offset)
```
[the remark this answers is not in the document]
Comment (Salmaafifi98) [reply src/tap/src/tapcell.i:109]: I removed all functions that are not used anymore. make_site_loc and build_row are used in ifp tcl code.

File: src/tap/src/tapcell.cpp (hunk 30 above)
Comment (Salmaafifi98) [reply src/tap/src/tapcell.cpp:1236]: What I understand it is a prefix then and the condition just checks if it's included in the name or not, should something be changed?

File: src/tap/src/tapcell.cpp [hunk 33]
```diff
@@ -33,41 +33,1233 @@
...
+    }
+    if (macro0 != min_max_x.second) {
+      // Row is cut at the end
+      std::pair<int, int> macro0_macro1(macro0, -1);
+      macro_outlines[macro0_macro1].push_back(cur_row);
+    }
+  }
+
+  std::map<std::pair<int, int>, vector<int>> macro_outlines_array;
+  for(const auto&[k,ignored]:macro_outlines) {
```
[the remark this answers is not in the document]
Comment (Salmaafifi98) [reply src/tap/src/tapcell.cpp:1056]: Yes, thank you!

File: src/tap/include/tap/tapcell.h [hunk 34]
```diff
@@ -51,118 +51,136 @@ class dbBlock;
...
-                int& halo_x,
-                int& halo_y);
+               const std::vector<odb::dbBox*>& blockages,
+               int halo_x,
+               int halo_y);
+  int removeCells(const std::string& prefix);
   std::vector<odb::dbBox*> findBlockages();
  
  private:
+  struct filled_sites
```
Comment (maliberty) [inline src/tap/include/tap/tapcell.h:105]: Should be FilledSites (see https://google.github.io/styleguide/cppguide.html#Type_Names)

File: src/tap/include/tap/tapcell.h [hunk 35]
```diff
@@ -51,118 +51,136 @@ class dbBlock;
...
+  {
+    int yMin;
+    int xMin;
+    int xMax;
+  };
+  struct cnrcap_masters
```
Comment (maliberty) [inline src/tap/include/tap/tapcell.h:111]: CnrcapMasters

What is cnrcap?

File: src/tap/src/tapcell.cpp [hunk 36]
```diff
@@ -2,7 +2,7 @@
 //
 // BSD 3-Clause License
 //
-// Copyright (c) 2019, The Regents of the University of California
+// Copyright (c) 2019, University of California, San Diego.
```
Comment (maliberty) [inline src/tap/src/tapcell.cpp:5]: Please don't change this

File: src/tap/src/tapcell.tcl [hunk 37]
```diff
@@ -2,7 +2,7 @@
 ##
 ## BSD 3-Clause License
 ##
-## Copyright (c) 2019, The Regents of the University of California
+## Copyright (c) 2019, University of California, San Diego.
```
Comment (maliberty) [inline src/tap/src/tapcell.tcl:5]: revert

**Review by maliberty — CHANGES_REQUESTED** [review]
Almost there

File: src/tap/include/tap/tapcell.h (hunk 35 above)
Comment (eder-matheus) [reply src/tap/include/tap/tapcell.h:111]: AFAIK, they are cells placed at the corners of macros and at the corners of the core area.

File: src/tap/src/tapcell.cpp (hunk 36 above)
Comment (Salmaafifi98) [reply src/tap/src/tapcell.cpp:5]: Sorry, I did not notice, they apparently get changed automatically whenever I pull, reverted now.

File: src/tap/include/tap/tapcell.h (hunk 35 above)
Comment (maliberty) [reply src/tap/include/tap/tapcell.h:111]: It sounds like cnr == corner

Comment (eder-matheus) [reply src/tap/include/tap/tapcell.h:111]: Agreed. The struct can be renamed CornerCapMaster.

**Comment by maliberty** [comment]
@Salmaafifi98 the CI tests are not passing, please take a look.

**Comment by Salmaafifi98** [comment]
> @Salmaafifi98 the CI tests are not passing, please take a look.

It is fixed now.

**Comment by maliberty** [comment]
@eder-matheus anything more here or shall I approve?

File: src/tap/src/tapcell.cpp (hunk 30 above)
Comment (maliberty) [reply src/tap/src/tapcell.cpp:1239]: "abc".find("b") will != npos because "b" is somewhere in the string, not just at the start (ie a prefix).

File: src/tap/src/tapcell.i [hunk 38]
```diff
@@ -34,14 +34,132 @@
...
+                int halo_y) {
+    return getTapcell()->overlaps(blockage, row, halo_x, halo_y);
+  }
+
+
+
+
+  }  // namespace tap
+
+  %}
```
Comment (eder-matheus) [inline src/tap/src/tapcell.i:165]: Add a new line after `%}`

File: src/tap/src/tapcell.cpp [hunk 39]
```diff
@@ -33,41 +33,1236 @@
...
+      return symmetry_y;
+    case odb::dbOrientType::R180:
+      return (symmetry_x && symmetry_y);
+    default:
+      return false;
+  }
+}
+
+// namespace end
+}  // namespace tap
```
Comment (eder-matheus) [inline src/tap/src/tapcell.cpp:1268]: Add a new line after `}  // namespace tap`

File: src/tap/src/MakeTapcell.cpp [hunk 40]
```diff
@@ -33,30 +33,44 @@
...
-			    openroad->getDb());
+  Tcl_Interp* tcl_interp = openroad->tclInterp();
+  Tap_Init(tcl_interp);
+  // Eval encoded sta TCL sources.
+  sta::evalTclInit(tcl_interp, sta::tap_tcl_inits);
+  openroad->getTapcell()->init(openroad->getDb(), openroad->getLogger());
 }
 
-}
+}  // namespace ord
```
Comment (eder-matheus) [inline src/tap/src/MakeTapcell.cpp:76]: Add a new line after `}  // namespace ord`

File: src/tap/include/tap/tapcell.h [hunk 41]
```diff
@@ -33,23 +33,155 @@
...
+                    odb::dbMaster* endcap_master,
+                    const CornercapMasters& masters);
+  std::vector<std::vector<odb::dbRow*>> organizeRows();
+  int insertTapcells(const std::vector<std::vector<odb::dbRow*>>& rows,
+                     const std::string& tapcell_master,
+                     int dist);
 };
 
-}
+}  // namespace tap
```
Comment (eder-matheus) [inline src/tap/include/tap/tapcell.h:187]: Add a new line after } // namespace tap

File: src/tap/include/tap/MakeTapcell.h [hunk 42]
```diff
@@ -45,14 +44,11 @@ namespace ord {
...
-deleteTapcell(tap::Tapcell *tapcell);
+void deleteTapcell(tap::Tapcell* tapcell);
 
-void
-initTapcell(OpenRoad *openroad);
+void initTapcell(OpenRoad* openroad);
 
-} // namespace
+}  // namespace ord
 #endif
```
Comment (eder-matheus) [inline src/tap/include/tap/MakeTapcell.h:54]: Add a new line after `#endif`

File: src/tap/include/tap/MakeTapcell.h [hunk 43]
```diff
@@ -33,7 +33,6 @@
 //
 ///////////////////////////////////////////////////////////////////////////////
 
-
 #ifndef MAKE_TAPCELL_H
 #define MAKE_TAPCELL_H
```
Comment (eder-matheus) [inline src/tap/include/tap/MakeTapcell.h:37]: Use `#pragma once` instead of `#ifndef MAKE_TAPCELL_H ` and `#define MAKE_TAPCELL_H `

File: src/ifp/src/InitFloorplan.tcl [hunk 44]
```diff
@@ -291,10 +328,10 @@ proc cut_rows {block placement_blockages} {
...
   foreach row $rows_to_cut {
-    tap::cut_row $block $row $row_placement_blockages 0 0 0
+    ifp::cut_row $block $row $row_placement_blockages 0 0 0
   }
 
   utl::info "IFP" 23 "Cut rows: [llength $rows_to_cut]"
 }
 
-}
+}
```
Comment (eder-matheus) [inline src/ifp/src/InitFloorplan.tcl:337]: Add a new line

**Comment by eder-matheus** [comment]
> @eder-matheus anything more here or shall I approve?

I've added a few more comments, but they are minor. After these and the prefix are fixed, we can approve.

File: src/tap/src/tapcell.cpp (hunk 30 above)
Comment (maliberty) [reply src/tap/src/tapcell.cpp:1239]: You can use == 0 to test that it is at the start rather than != string::npos

File: src/tap/include/tap/MakeTapcell.h (hunk 42 above)
Comment (Salmaafifi98) [reply src/tap/include/tap/MakeTapcell.h:54]: There are already one new line after #endif and some of the other files as well. Do I add another or should there just be one more line at the end of each file?

**Review by maliberty — APPROVED** [review]
